compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3851,63 +3860,193 @@ object Types {
     def apply(tp: Type) = tp
   }
 
-  /** A type map that approximates NoTypes by upper or lower known bounds depending on
+  /** A type map that approximates TypeBounds types depending on
    *  variance.
    *
    *  if variance > 0 : approximate by upper bound
    *     variance < 0 : approximate by lower bound
-   *     variance = 0 : propagate NoType to next outer level
+   *     variance = 0 : propagate bounds to next outer level
    */
   abstract class ApproximatingTypeMap(implicit ctx: Context) extends TypeMap { thisMap =>
-    def approx(lo: Type = defn.NothingType, hi: Type = defn.AnyType) =
-      if (variance == 0) NoType
-      else apply(if (variance < 0) lo else hi)
 
+    protected def range(lo: Type, hi: Type) =
+      if (variance > 0) hi
+      else if (variance < 0) lo
+      else Range(lower(lo), upper(hi))
+
+    protected def isRange(tp: Type) = tp.isInstanceOf[Range]
+
+    protected def lower(tp: Type) = tp match {
+      case tp: Range => tp.lo
+      case _ => tp
+    }
+
+    protected def upper(tp: Type) = tp match {
+      case tp: Range => tp.hi
+      case _ => tp
+    }
+
+    protected def rangeToBounds(tp: Type) = tp match {
+      case Range(lo, hi) => TypeBounds(lo, hi)
+      case _ => tp
+    }
+
+    protected def atVariance[T](v: Int)(op: => T): T = {
```

Comment:
I think this is worth putting in `TypeMap` itself, considering how common the pattern `val saved = variance; variance = ...` is. Also maybe worth marking `inline` ?